{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def ensure_dirs_exists(path):\n",
    "    if \".\" in path:\n",
    "        path = os.path.dirname(path)\n",
    "    if not os.path.exists(path):\n",
    "        os.makedirs(path)\n",
    "    return\n",
    "\n",
    "def writeToFile(filename, dataframe):\n",
    "    write = filename + '_Transformed.csv'\n",
    "    dataframe.to_csv(write, index=True)\n",
    "    print(\"file saved to: {}\".format(write))\n",
    "    return\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Accession', 'Description', 'Genes', 'Peptides', 'Detected_Imputed',\n",
      "       '110129 Homo_H Normalized', '110130 Homo_H Normalized',\n",
      "       '110131 Homo_H Normalized', '110132 Super_S1 Normalized',\n",
      "       '110133 Super_S1 Normalized', '110134 Super_S1 Normalized',\n",
      "       '110135 CrudeSyn_P2 Normalized', '110136 CrudeSyn_P2 Normalized',\n",
      "       '110137 CrudeSyn_P2 Normalized', '110138 Syn_LP1 Normalized',\n",
      "       '110139 Syn_LP1 Normalized', '110140 Syn_LP1 Normalized'],\n",
      "      dtype='object') \n",
      " ['110160_01 SPQC Normalized', '110160_02 SPQC Normalized', '110160_03 SPQC Normalized', 'Homo_H_CV', 'Super_S1_CV', 'CrudeSyn_P2_CV', 'Syn_LP1_CV', 'SPQC_CV']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein</th>\n",
       "      <th>Origin</th>\n",
       "      <th>Mixture</th>\n",
       "      <th>Function</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>BioFraction</th>\n",
       "      <th>Description</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Peptides</th>\n",
       "      <th>Detected_Imputed</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Intensity</th>\n",
       "      <th>Total_Intensity</th>\n",
       "      <th>Rel_Intensity</th>\n",
       "      <th>Abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A087WPF7</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>H</td>\n",
       "      <td>Autism susceptibility gene 2 protein homolog</td>\n",
       "      <td>WildType</td>\n",
       "      <td>r1</td>\n",
       "      <td>Autism susceptibility gene 2 protein homolog, ...</td>\n",
       "      <td>Auts2</td>\n",
       "      <td>2</td>\n",
       "      <td>2.2.2.2.2.2.1.2.-.-.1.1.1.2.2</td>\n",
       "      <td>110129 Homo_H Normalized</td>\n",
       "      <td>543.118684</td>\n",
       "      <td>3057.169232</td>\n",
       "      <td>0.177654</td>\n",
       "      <td>9.085124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A087WPF7</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>H</td>\n",
       "      <td>Autism susceptibility gene 2 protein homolog</td>\n",
       "      <td>WildType</td>\n",
       "      <td>r2</td>\n",
       "      <td>Autism susceptibility gene 2 protein homolog, ...</td>\n",
       "      <td>Auts2</td>\n",
       "      <td>2</td>\n",
       "      <td>2.2.2.2.2.2.1.2.-.-.1.1.1.2.2</td>\n",
       "      <td>110130 Homo_H Normalized</td>\n",
       "      <td>546.194965</td>\n",
       "      <td>3057.169232</td>\n",
       "      <td>0.178660</td>\n",
       "      <td>9.093272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A087WPF7</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>H</td>\n",
       "      <td>Autism susceptibility gene 2 protein homolog</td>\n",
       "      <td>WildType</td>\n",
       "      <td>r3</td>\n",
       "      <td>Autism susceptibility gene 2 protein homolog, ...</td>\n",
       "      <td>Auts2</td>\n",
       "      <td>2</td>\n",
       "      <td>2.2.2.2.2.2.1.2.-.-.1.1.1.2.2</td>\n",
       "      <td>110131 Homo_H Normalized</td>\n",
       "      <td>239.267457</td>\n",
       "      <td>3057.169232</td>\n",
       "      <td>0.078264</td>\n",
       "      <td>7.902480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A087WPF7</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>S1</td>\n",
       "      <td>Autism susceptibility gene 2 protein homolog</td>\n",
       "      <td>WildType</td>\n",
       "      <td>r1</td>\n",
       "      <td>Autism susceptibility gene 2 protein homolog, ...</td>\n",
       "      <td>Auts2</td>\n",
       "      <td>2</td>\n",
       "      <td>2.2.2.2.2.2.1.2.-.-.1.1.1.2.2</td>\n",
       "      <td>110132 Super_S1 Normalized</td>\n",
       "      <td>495.669246</td>\n",
       "      <td>3057.169232</td>\n",
       "      <td>0.162133</td>\n",
       "      <td>8.953234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A087WPF7</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>S1</td>\n",
       "      <td>Autism susceptibility gene 2 protein homolog</td>\n",
       "      <td>WildType</td>\n",
       "      <td>r2</td>\n",
       "      <td>Autism susceptibility gene 2 protein homolog, ...</td>\n",
       "      <td>Auts2</td>\n",
       "      <td>2</td>\n",
       "      <td>2.2.2.2.2.2.1.2.-.-.1.1.1.2.2</td>\n",
       "      <td>110133 Super_S1 Normalized</td>\n",
       "      <td>249.361333</td>\n",
       "      <td>3057.169232</td>\n",
       "      <td>0.081566</td>\n",
       "      <td>7.962094</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Protein        Origin Mixture  \\\n",
       "0  A0A087WPF7  Mus musculus      H    \n",
       "1  A0A087WPF7  Mus musculus      H    \n",
       "2  A0A087WPF7  Mus musculus      H    \n",
       "3  A0A087WPF7  Mus musculus     S1    \n",
       "4  A0A087WPF7  Mus musculus     S1    \n",
       "\n",
       "                                       Function  Genotype BioFraction  \\\n",
       "0  Autism susceptibility gene 2 protein homolog  WildType          r1   \n",
       "1  Autism susceptibility gene 2 protein homolog  WildType          r2   \n",
       "2  Autism susceptibility gene 2 protein homolog  WildType          r3   \n",
       "3  Autism susceptibility gene 2 protein homolog  WildType          r1   \n",
       "4  Autism susceptibility gene 2 protein homolog  WildType          r2   \n",
       "\n",
       "                                         Description   Gene  Peptides  \\\n",
       "0  Autism susceptibility gene 2 protein homolog, ...  Auts2         2   \n",
       "1  Autism susceptibility gene 2 protein homolog, ...  Auts2         2   \n",
       "2  Autism susceptibility gene 2 protein homolog, ...  Auts2         2   \n",
       "3  Autism susceptibility gene 2 protein homolog, ...  Auts2         2   \n",
       "4  Autism susceptibility gene 2 protein homolog, ...  Auts2         2   \n",
       "\n",
       "                Detected_Imputed                      Sample   Intensity  \\\n",
       "0  2.2.2.2.2.2.1.2.-.-.1.1.1.2.2    110129 Homo_H Normalized  543.118684   \n",
       "1  2.2.2.2.2.2.1.2.-.-.1.1.1.2.2    110130 Homo_H Normalized  546.194965   \n",
       "2  2.2.2.2.2.2.1.2.-.-.1.1.1.2.2    110131 Homo_H Normalized  239.267457   \n",
       "3  2.2.2.2.2.2.1.2.-.-.1.1.1.2.2  110132 Super_S1 Normalized  495.669246   \n",
       "4  2.2.2.2.2.2.1.2.-.-.1.1.1.2.2  110133 Super_S1 Normalized  249.361333   \n",
       "\n",
       "   Total_Intensity  Rel_Intensity  Abundance  \n",
       "0      3057.169232       0.177654   9.085124  \n",
       "1      3057.169232       0.178660   9.093272  \n",
       "2      3057.169232       0.078264   7.902480  \n",
       "3      3057.169232       0.162133   8.953234  \n",
       "4      3057.169232       0.081566   7.962094  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def transform_raw(filename, sheetname):\n",
    "    if \"xlsx\" in filename:\n",
    "        df = pd.read_excel(filename, sheet_name= sheetname)\n",
    "    if \"csv\" in filename:\n",
    "        df = pd.read_csv(filename)\n",
    "    pd.set_option('display.max_rows', None)\n",
    "    colsdrop = [x for x in df.columns if '_CV' in x.upper() or 'SPQC' in x.upper()]\n",
    "    df = df.drop(columns=colsdrop)\n",
    "    \n",
    "    print(df.columns, '\\n', colsdrop)\n",
    "    H = list(range(5, 8))\n",
    "    S1 = list(range(8, 11))\n",
    "    P2 = list(range(11, 14))\n",
    "    LP1 = list(range(14, 17))\n",
    "    long_df = pd.melt(df, \n",
    "            id_vars=['Accession', 'Description', 'Genes', 'Peptides', 'Detected_Imputed'],\n",
    "            var_name='Sample', \n",
    "            value_name='Intensity') # raw intensity\n",
    "    abundance_sum = long_df.groupby('Accession')['Intensity'].sum().reset_index(name='Total_Intensity')\n",
    "\n",
    "    # Merge this sum back into the long dataframe\n",
    "    long_df = long_df.merge(abundance_sum, on='Accession')\n",
    "    long_df['Rel_Intensity'] = long_df['Intensity'] / long_df['Total_Intensity']\n",
    "    long_df['Abundance']= np.log2(long_df['Intensity'])\n",
    "    long_df.insert(1, 'Origin', long_df['Description'].str.split(\"org=\").str.get(-1))\n",
    "    long_df.insert(2, 'Mixture', long_df['Sample'].str.split(\"_\").str.get(-1).str.split(\"Normalized\").str.get(0))\n",
    "    long_df.insert(3, 'Function', long_df['Description'].str.split(\",\").str.get(0))\n",
    "    long_df.insert(4, 'Genotype', 'WildType')\n",
    "    long_df.insert(5, 'BioFraction', long_df.groupby(['Accession', 'Mixture']).cumcount() + 1)\n",
    "    long_df['BioFraction'] = 'r' + long_df['BioFraction'].astype(str)\n",
    "    # long_df.drop(columns= ['Description'], inplace=True)\n",
    "    long_df.rename(columns={'Accession': 'Protein',\n",
    "                            'Genes': 'Gene'}, inplace=True)\n",
    "    return long_df\n",
    "filename = 'data_10297JBK/10297_SupplementalData_Proteome_122723.xlsx'\n",
    "worksheet = 'Normalized Data'\n",
    "\n",
    "currdir = os.getcwd()\n",
    "parent = os.path.dirname(currdir)\n",
    "gparent = os.path.dirname(os.path.dirname(currdir))\n",
    "rawdatafile = f\"{gparent}/example_data/rawdata/{filename}\" # EXAMPLE DATA DIR, change for you. \n",
    "# do conversion\n",
    "dataframe = transform_raw(rawdatafile, worksheet)\n",
    "\n",
    "dataframe.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Accession', 'Description', 'Genes', 'Peptides', 'Detected_Imputed',\n",
      "       '110129 Homo_H Normalized', '110130 Homo_H Normalized',\n",
      "       '110131 Homo_H Normalized', '110132 Super_S1 Normalized',\n",
      "       '110133 Super_S1 Normalized', '110134 Super_S1 Normalized',\n",
      "       '110135 CrudeSyn_P2 Normalized', '110136 CrudeSyn_P2 Normalized',\n",
      "       '110137 CrudeSyn_P2 Normalized', '110138 Syn_LP1 Normalized',\n",
      "       '110139 Syn_LP1 Normalized', '110140 Syn_LP1 Normalized'],\n",
      "      dtype='object') \n",
      " ['110160_01 SPQC Normalized', '110160_02 SPQC Normalized', '110160_03 SPQC Normalized', 'Homo_H_CV', 'Super_S1_CV', 'CrudeSyn_P2_CV', 'Syn_LP1_CV', 'SPQC_CV']\n",
      "file saved to: /home/poojaparameswaran/Documents/SoderlingLab/SpatialProteomics/transformeddata/data_10297JBK/10297_SupplementalData_Proteome_122723_Transformed.csv\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def main():\n",
    "    ## SNCA == Normalized Data, LRRK2 == Table 4 Normalized Data\n",
    "    ## Vps35 == Table 3 Normalized Data\n",
    "    filename = 'data_10297JBK/10297_SupplementalData_Proteome_122723.xlsx'\n",
    "    worksheet = 'Normalized Data'\n",
    "    \n",
    "    currdir = os.getcwd()\n",
    "    parent = os.path.dirname(currdir)\n",
    "    gparent = os.path.dirname(os.path.dirname(currdir))\n",
    "    rawdatafile = f\"{gparent}/example_data/rawdata/{filename}\" # EXAMPLE DATA DIR, change for you. \n",
    "    # do conversion\n",
    "    dataframe = transform_raw(rawdatafile, worksheet)\n",
    "\n",
    "    # write to file\n",
    "    filename = filename.replace(\".xlsx\", \"\")\n",
    "    newfilepath =  os.path.join(gparent, \"transformeddata\", filename)\n",
    "    ensure_dirs_exists(newfilepath)\n",
    "    writeToFile(newfilepath, dataframe)\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dataenv",
   "language": "python",
   "name": "dataenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
